Document building against the Microsoft.NETFramework.ReferenceAssemblies NuGet package (#32762)
dotnet/docs: docs/framework/migration-guide/reference-assemblies.md

---
title: "Build apps against Microsoft.NETFramework.ReferenceAssemblies"
description: "Developers can target .NET Framework by referencing the official reference assemblies NuGet package."
ms.date: 11/25/2022
helpviewer_keywords:
  - ".NET Framework, targeting"
---
# Build apps against Microsoft.NETFramework.ReferenceAssemblies

When you target a particular version of .NET Framework, by default your application is built by using the reference assemblies that are included with that version's developer pack. In scenarios where the matching developer pack cannot be installed on the computer, you can alternatively build against reference assemblies distributed via a NuGet package.

## Update project files

Each project that should build against the reference assemblies NuGet package needs to include the following package reference.

```xml
<ItemGroup>
  <PackageReference Include="Microsoft.NETFramework.ReferenceAssemblies" Version="1.0.3" PrivateAssets="All" />
</ItemGroup>

> ```

Author: Non-SDK supports package references in project files as well:
https://learn.microsoft.com/en-us/nuget/consume-packages/package-references-in-project-files#project-type-support

I believe that the default in VS is still packages.config for compatibility reasons but can be flipped in options somewhere. Let me check if one project can use both (existing packages.config + `PackageReference`). If not, I guess I'll have to document both ways here.

> ```

Author: Turns out NuGet doesn't support mix-and-match. The project is either `packages.config`-based or `PackageReference`-based. The former has poor command-line support so I think it's better to recommend only the VS Package Manager to add the reference. Oh, this is so convoluted :)

> ```

Author: I've added a `packages.config` snippet as you suggested.

> SDK-style projects include this reference by default. For traditional, non-SDK-style projects, the reference can be added with the NuGet Package Manager UI in Visual Studio or by manually editing the project file. The package contains reference assemblies for many versions of .NET Framework. The version to be actually used is determined by the TargetFramework (TargetFrameworks) property.

Author: Thank you, the suggested change is awesome. I copied the "traditional" from https://github.com/dotnet/docs/blame/main/docs/csharp/language-reference/preprocessor-directives.md#L116. On second thought I agree it is not appropriate.

Great catch on `TargetFramework`. All add `TargetFrameworkVersion` and qualify with SDK / non-SDK.

> [!NOTE]

> > Projects that contain a project reference need to be restored before they can be built. When building with Visual Studio, either in the IDE or on the command line with *MSBuild.exe*, make sure to explicitly execute the *restore* target prior to building the project. This is done on the Developer Command Prompt by executing `msbuild /t:restore` in the project directory.

Reviewer: Avoid the work "execute" if you can
```suggestion
> Projects that contain a project reference need to be restored before they can be built. When building with Visual Studio, either in the IDE or on the command line with *MSBuild.exe*, make sure to explicitly run the *restore* target prior to building the project. This is done on the Developer Command Prompt with the `msbuild /t:restore` command from the project directory.
```

> > Projects that contain a project reference need to be restored before they can be built. When building with Visual Studio, either in the IDE or on the command line with *MSBuild.exe*, make sure to explicitly execute the *restore* target prior to building the project. This is done on the Developer Command Prompt by executing `msbuild /t:restore` in the project directory.

Reviewer: Also, running `msbuild /t:restore` is really something someone would run if they were using Framework-style projects and not SDK-style projects. So along with my other suggestion of including instructions for both project styles, I would update this specifically to be `dotnet restore` and then use `msbuild` in the Framework-style section.

Author: This guidance is primarily for non-SDK-style ("Framework-style" as you wrote) projects. SDK-style works out of the box because they already have the `PackageReference` included from the .NET SDK.

SDK-style still needs to run restore, but that's something they have to do anyways, so it looks as though including it here would be redundant. Although on second thought, it makes sense that SDK-style devs would land here too, and the step to restore may not be obvious, especially since "Restore NuGet Packages" in VS doesn't cut it.

OK, let me restructure it a bit. Thank you!

Author: I ran through the scenarios again and have updated the text.
- The build prefers the machine-wide targeting pack in all cases. So I have removed the "non-SDK-style" part from the first paragraph. The new text is more correct and concise.
- The section about updating project files explains the difference between SDK and non-SDK. It doesn't recommend `dotnet add-package` anymore, because it's not needed for SDK and doesn't work for non-SDK.
- A new section explains how to restore on the command line. The difference here is not between SDK and non-SDK, but between VS and dotnet. While likely equivalent in most cases, I think it's better to recommend `msbuild /t:restore` even for SDK-style projects if the developer is in VS. @zivkan, could you please comment on this? Thank you!

Reviewer 2: Some (most? all?) .NET Framework (non-SDK) projects can't be built, or restored, by the `dotnet` CLI. The project files import props/targets that only ship in VS. However, VS and msbuild is able to discover the .NET SDK. So, `msbuild -t:restore` will work for anyone with VS installed, but `dotnet restore` will only work for solutions where every* project in the solution is SDK style. Just one non-SDK style project (like a Visual Studio Extension VSIX project) will cause `dotnet` to fail MSBuild evaluation.

So, while I personally do recommend customers with all SDK style projects in their solution to use the dotnet cli (just so the skills are transferrable to Linux & Mac, no technical reason), it is safe to tell anyone on Windows with VS installed to use msbuild.

*I think that very basic console and classlib projects might work with the dotnet cli, but it's just easier to say something like "non-SDK style project, you must use msbuild".

Author: Thank you. The text now says "When building with VS ... use `nuget restore` or `msbuild /t:restore`" and "When building with .NET SDK ... use `dotnet restore`"). That should be safe / work for everyone, unless I misunderstood.

---
title: "Build apps against Microsoft.NETFramework.ReferenceAssemblies"

> description: "Developers can target .NET Framework by referencing the official reference assemblies NuGet package."
> ms.date: 11/25/2022

Reviewer: I altered the description a bit. I think this is more search engine friendly. I also updated the article date 😁 
```suggestion
description: "Learn how to reference the .NET Framework Reference Assemblies in your project, when you can't install a matching developer pack."
ms.date: 01/19/2023
```

Author: 👍

Each project that should build against the reference assemblies NuGet package needs to include a reference to _Microsoft.NETFramework.ReferenceAssemblies_.

Projects using [_packages.config_](/nuget/reference/packages-config) should include the following in _packages.config_.

```xml
<packages>
  <package id="Microsoft.NETFramework.ReferenceAssemblies" version="1.0.3" developmentDependency="true" />
</packages>
```

Projects using [`<PackageReference>`](/nuget/consume-packages/package-references-in-project-files) should include the following in the project file.

```xml
<ItemGroup>
  <PackageReference Include="Microsoft.NETFramework.ReferenceAssemblies" Version="1.0.3" PrivateAssets="All" />
</ItemGroup>
```

SDK-style projects include this reference by default. For typical .NET Framework projects that were created with Visual Studio, the reference can be added with the NuGet Package Manager UI in Visual Studio. The package contains reference assemblies for many versions of .NET Framework. The version to be actually used is determined by the `TargetFrameworkVersion` or `TargetFramework` (`TargetFrameworks`) property, as already defined in the project file.

> ## Run restore on the command line
> 
> Projects that contain a package reference need to be restored before they can be built.
> 
> When building with Visual Studio, either in the IDE or on the command line with *MSBuild.exe*, make sure to explicitly restore packages prior to building the project. For projects using _packages.config_, this done with [nuget restore](/nuget/reference/cli-reference/cli-ref-restore). For projects using `<PackageReference>`, this is done on the Developer Command Prompt with the `msbuild /t:restore` command from the project directory.
> 
> When building with the .NET SDK, make sure to run the [dotnet restore](../../core/tools/dotnet-restore.md) command from the project directory.
> 
> ## References
> 
> - [.NET Framework Targeting Pack Nuget Packages](https://github.com/microsoft/dotnet/tree/main/releases/reference-assemblies)

Reviewer: What do you think of this change? This rewrites the section to be explicit in the steps someone needs to take to restore their project, which from your description, seems to be a key to getting it to work.

I also deleted the **## References** section, as we don't really those anymore.

I also changed the .NET SDK text, since .NET Core 2.0, a restore action happens every time the project is run or compiled.

```suggestion

## Restore the project

Projects that contain a package reference must be restored before they can be built.

After adding the **Microsoft.NETFramework.ReferenceAssemblie** NuGet package to your project, you must explicitly run the restore action in one of the following ways:

- If your project is a non SDK-style project and uses the _packages.config_ file to reference NuGet packages:

  01. Install the [NuGet CLI tool](https://learn.microsoft.com/en-us/nuget/install-nuget-client-tools#nugetexe-cli), making sure _nuget.exe_ is in the PATH variable.
  01. Open a command prompt.
  01. Navigate to the directory that contains your project file.
  01. Run _nuget.exe restore_.

- If your project is a non SDK-style project and uses `<PackageReference>` settings in the project file to reference NuGet packages:

  01. Open **Developer Command Prompt for VS 2022**. The name of this app may be different based on which version of Visual Studio you've installed.
  01. Navigate to the directory that contains your project file.
  01. Run _msbuild /t:restore_.

- If your project is an SDK-style project, you don't need to do anything. The NuGet restore action is automatically run when the project is built.
```

Author: Much better, thank you!

(I have edited your suggestion to fix a typo `Microsoft.NETFramework.ReferenceAssemblie` -> `Microsoft.NETFramework.ReferenceAssemblies` before committing.)
